File: diagonalTraverse.py
from typing import List

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if not mat or not mat[0]:
            return []
        
        m, n = len(mat), len(mat[0])
        result = []
        for d in range(m + n - 1):
            intermediate = []
            
            # Determine the starting point
            r = 0 if d < n else d - n + 1
            c = d if d < n else n - 1
            
            # Collect current diagonal
            while r < m and c >= 0:
                intermediate.append(mat[r][c])
                r += 1
                c -= 1
            
            # Reverse every other diagonal
            if d % 2 == 0:
                result.extend(intermediate[::-1])
            else:
                result.extend(intermediate)
        
        return result

File: diagonalTraverse.swift
class Solution {
    func findDiagonalOrder(_ mat: [[Int]]) -> [Int] {
        let m = mat.count
        let n = mat[0].count
        var result: [Int] = []

        for d in 0..<(m + n - 1) {
            var intermediate: [Int] = []

            var r = d < n ? 0 : d - n + 1
            var c = d < n ? d : n - 1

            while r < m && c >= 0 {
                intermediate.append(mat[r][c])
                r += 1
                c -= 1
            }

            if d % 2 == 0 {
                result.append(contentsOf: intermediate.reversed())
            } else {
                result.append(contentsOf: intermediate)
            }
        }

        return result
    }
}
